overlord/snapstate/snapstate_test.go

```diff
@@ -9229,6 +9322,42 @@ type testResolveValidationSetsEnforcementErrorComponentsOpts struct {
 	current           []snapstate.SnapState
 }
 
+func validateEnforcementOrder(c *C, st *state.State, tss []*state.TaskSet) {
+	deviceCtx, err := snapstate.DeviceCtxFromState(st, nil)
+	c.Assert(err, IsNil)
+
+	essentials := []string{"snapd"}
+	for _, sn := range deviceCtx.Model().EssentialSnaps() {
+		essentials = append(essentials, sn.SnapName())
+	}
+
+	var essentialTasks []string
+	for _, ts := range tss {
+		begin := ts.MaybeEdge(snapstate.BeginEdge)
+		if begin == nil {
+			continue
+		}
+
+		snapsup, err := snapstate.TaskSnapSetup(begin)
+		c.Assert(err, IsNil)
+
+		switch {
+		case strutil.ListContains(essentials, snapsup.SideInfo.RealName):
+			// essential snap updates don't wait on anything
+			c.Assert(begin.WaitTasks(), HasLen, 0)
+			for _, t := range ts.Tasks() {
+				essentialTasks = append(essentialTasks, t.ID())
+			}
+		case snapsup.Type == snap.TypeBase:
+			// non-essential bases only should wait on tasks that come from
+			// essential snap updates
+			for _, t := range begin.WaitTasks() {
+				c.Assert(strutil.ListContains(essentialTasks, t.ID()), Equals, true)
+			}
+		}
```

Comment:
Should there be a check that the non-essential and non-base tasks wait for the updates and installs of bases?